keras_cv/models/segmentation/fcn.py

```diff
@@ -0,0 +1,646 @@
+        "DENSE_UNITS": [4096, 4096],
+    },
+    "vgg19": {
+        "BLOCK3": "block3_pool",
+        "BLOCK4": "block4_pool",
+        "BLOCK5": "block5_pool",
+        "DENSE_UNITS": [4096, 4096],
+    },
+}
+BACKBONE = {"vgg16": VGG16, "vgg19": VGG19}
+
+
+def get_dense_to_convolution_layers(model):
+    """Utility function to convert all `tf.keras.layers.Dense` functions to `tf.keras.layers.Conv2D` with `filters=units` from the Dense layers and `kernel_size=(1,1)`, `padding='same'`
+
+    Args:
+        model: `tf.keras.layers.Model` or `tf.keras.Sequential`, The original model from which the layers are extracted
+    Returns:
+        layers: `tf.keras.Sequential` representing the converted layers.
+    """
+
+    if model.name.lower() == "vgg16":
+        backbone_name = "vgg16"
+    elif model.name.lower() == "vgg19":
+        backbone_name = "vgg19"
+
+    units = BACKBONE_CONFIG[backbone_name]["DENSE_UNITS"]
+    dense_convs = []
+
+    for filter_idx in range(len(units)):
+        filter = units[filter_idx]
+        if filter_idx == 0:
+            dense_conv = tf.keras.layers.Conv2D(
+                filters=filter,
+                kernel_size=(
+                    7,
+                    7,
+                ),
+                strides=(1, 1),
+                activation="relu",
+                padding="same",
+            )
+            dense_convs.append(dense_conv)
+        else:
+            dense_conv = tf.keras.layers.Conv2D(
+                filters=filter,
+                kernel_size=(1, 1),
+                strides=(1, 1),
+                activation="relu",
+                padding="same",
+            )
+            dense_convs.append(dense_conv)
+        dropout_layer = tf.keras.layers.Dropout(0.5)
+        dense_convs.append(dropout_layer)
+    return tf.keras.Sequential(dense_convs)
+
+
+def vgg_backbone_builder(
+    backbone, model_architecture, input_shape=(None, None, 3), input_tensor=None
+):
+    """Utility function to build the backbone of the FCN with variants of the VGG model.
+
+    Args:
+        backbone_name: str, one of 'vgg16' or 'vgg19'. Defines which model to use as the backbone base.
+        model_architecture: str, one of 'fcn8s', 'fcn16s' or 'fcn32s'.
+            Defines which architecture and sampling method should be used as detailed in the
+            paper [Fully Convolutional Networks for Semantic Segmentation](https://arxiv.org/pdf/1411.4038.pdf)
+        input_shape: `list` or `tuple`. Defines the shape of the tensor to be expected as input.
+
+    Returns:
+        model: tf.keras.models.Model. Represents the graph of all backbone operations and outputs for the chosen architecture.
+    """
+    vgg_model = backbone
+    if vgg_model.name.lower() == "vgg16":
+        backbone_name = "vgg16"
+    else:
+        backbone_name = "vgg19"
+
+    x = utils.parse_model_inputs(input_shape=input_shape, input_tensor=input_tensor)
+
+    if model_architecture == "fcn8s":
```

Comment:
Still has large sections handled by `if` statements.
Why not move this entire logic into the `FCN8S`, `FCN16S` and `FCN32S` methods?

Reply:
If we do that, a generalized `FullyConvolutionalNetwork` won't be possible, as they do differ a bit. If that is acceptable, I'll move it all into those functions. Would reduce a large section of code and simplify things too.

Reply:
Also, to note, doing so will mean that Functional Subclassing won't be happening. If we wish to perform Func. subclassing, then each of those individual architectures will be a separate `tf.keras.Model` definition (which is easily doable, but would then be better off being made in separate files)

Reply:
Since we've been going a bit back and forth on `if` statements - there might be something I'm missing.
Let me re-read the paper and implementation and I'll revisit this